File: Practica_1/src/gestorAplicacion/Evento.java
package gestorAplicacion;

import java.util.ArrayList;

/*
 * Clase Evento encargada de crear los eventos, los cuales son un tipo de servicio
 * que se le ofrece a los clientes, estos eventos pueden ser de diferentes tipos,
 * como por ejemplo: bodas, cumpleaños, fiestas, etc.
 * El evento puede usar diferentes servicios externos, como por ejemplo:
 * Sonido, Entretenimiento, etc.
 * El evento también puede realizarse en distintos lugares del Hotel:
 * Piscina, Salón, Terraza.
 * 
 */
public class Evento extends Servicio {
    // ATRIBUTOS
    private Lugar lugar;
    private Usuario cliente;
    private ArrayList<ServicioExterno> servicios;
    private String fecha;
    private int duracion;
    private int numeroAsistentes;
    private ArrayList<Empleado> empleados;

    // CONSTRUCTOR
    public Evento(Lugar lugar, Usuario cliente, ArrayList<ServicioExterno> servicios, String fecha, int duracion,
            int numeroAsistentes, ArrayList<Empleado> empleados) {
        super("Evento en " + lugar, calcularValor(lugar, servicios, duracion, numeroAsistentes));
        this.lugar = lugar;
        this.cliente = cliente;
        this.servicios = servicios;
        this.duracion = duracion;
        this.numeroAsistentes = numeroAsistentes;
        this.empleados = empleados;
    }

    // Métodos relevantes -----------------------------------------------------
    /*
     * Este método se encarga de calcular el valor monetario de un evento.
     * Este valor lo calcula sumando el valor del lugar en el que se realizará, el
     * valor individual
     * de cada uno de los servicios externos que se contrataron, la duración del
     * evento
     * y los asistentes a este.
     */
    public static int calcularValor(Lugar lugar, ArrayList<ServicioExterno> servicios, int duracion, int numeroAsistentes) {
        int valor = 0;
        valor += lugar.getValor();
        for (ServicioExterno servicioExterno : servicios) {
            valor += servicioExterno.getValor();
        }
        valor += duracion * 10000;
        valor += numeroAsistentes * 10000;
        return valor;
    }

    // -----------------------------------------------
    // GETTERS AND SETTERS
    public Lugar getLugar() {
        return lugar;
    }

    public void setLugar(Lugar lugar) {
        this.lugar = lugar;
        this.setNombre("Evento en " + lugar.getNombre());
    }

    public Usuario getCliente() {
        return cliente;
    }

    public void setCliente(Usuario cliente) {
        this.cliente = cliente;
        this.setNombre("Evento de " + cliente.getNombre());
    }

    public ArrayList<ServicioExterno> getServicios() {
        return servicios;
    }

    public void setServicios(ArrayList<ServicioExterno> servicios) {
        this.servicios = servicios;
    }

    // getter fecha
    public String getFecha() {
        return fecha;
    }

    public void setFecha(String fecha) {
        this.fecha = fecha;
    }

    public int getDuracion() {
        return duracion;
    }

    public void setDuracion(int duracion) {
        this.duracion = duracion;
    }

    public int getNumeroAsistentes() {
        return numeroAsistentes;
    }

    public void setNumeroAsistentes(int numeroAsistentes) {
        this.numeroAsistentes = numeroAsistentes;
    }
    
    public ArrayList<Empleado> getEmpleados() {
        return empleados;
    }

    public void setEmpleados(ArrayList<Empleado> empleados) {
        this.empleados = empleados;
    }

    // otros getters no tan obvios
    /*
     * Este método se encarga de retornar un ArrayList con los nombres de cada una
     * de las
     * empresas externas cuyos servicios fueron contratados para el evento.
     */
    public ArrayList<String> getEmpresasContratadas() {
        ArrayList<String> lista = new ArrayList<String>();

        for (ServicioExterno servicio : servicios) {
            lista.add(servicio.getEmpresaContratada());
        }

        return lista;
    }
}

File: Python/gestorAplicacion/Evento.py

"""
Clase Evento encargada de crear los eventos, los cuales son un tipo de servicio
que se le ofrece a los clientes, estos eventos pueden ser de diferentes tipos,
como por ejemplo: bodas, cumpleaños, fiestas, etc.
El evento puede usar diferentes servicios externos, como por ejemplo:
Sonido, Entretenimiento, etc.
El evento también puede realizarse en distintos lugares del Hotel:
Piscina, Salón, Terraza.
"""

class Evento(Servicio):
    
    """
    Constructor de la clase Evento:
    Parámetros:
        lugar: Lugar donde se realizará el evento.
        cliente: Cliente (objeto de tipo Usuario) que solicita el evento.
        servicios: Lista de servicios externos (objetos de tipo ServicioExterno que se usarán en el evento.
        fecha: Fecha en la que se realizará el evento en formato (dd/mm/aaaa).
        duracion: Duración del evento en minutos.
        numeroAsistentes: Número de asistentes al evento.
        empleados: Lista de empleados (objetos de clase Empleado) necesarios para el evento.
        codigo: Código del evento.
        descripcion (opcional): Descripción del evento.
    """
    def __init__(self, codigo, lugar, cliente, servicios, fecha, duracion, numeroAsistentes, empleados,descripcion=""):
        super().__init__(f"Evento en {lugar}" , self.calcularValor(lugar, servicios, duracion, numeroAsistentes, empleados))
        self.lugar = lugar
        self.cliente = cliente
        self.servicios = servicios
        self.duracion = duracion
        self.numeroAsistentes = numeroAsistentes
        self.empleados = empleados
        self.fecha = fecha
        self.codigo = codigo
        self.descripcion = descripcion
        
    
    """
    Método para calcular el valor monetario de evento en función
    de los servicios externos, el lugar, la duración, el número de asistentes.
    """
    @staticmethod
    def calcularValor(lugar, servicios, duracion, numeroAsistentes, empleados):
        valor = 0
        valor += lugar.getValor()
        for servicioExterno in servicios:
            valor += servicioExterno.getValor()
        valor += duracion * 10000
        valor += numeroAsistentes * 10000
        for empleado in empleados:
            valor += empleado.getNomina()
        return valor
    
    
    """
    ToString de la clase Evento.
    """
    def __str__(self):
        return "Lugar del evento: " + str(self.getLugar()) + "\n" + "Cliente asociado: " + str(self.getCliente()) + "\n" + "Fecha: " + str(self.getFecha()) + "\n" + "Duracion: " + str(self.getDuracion()) + "Numero de asistentes: " + str(self.getNumeroAsistentes())
    
    
    #Getters y Setters
    
    def getLugar(self):
        return self.lugar
    
    def setLugar(self, lugar):
        self.lugar = lugar
        
    def getCliente(self):
        return self.cliente
    
    def setCliente(self, cliente):
        self.cliente = cliente
        
    def getServicios(self):
        return self.servicios
    
    def setServicios(self, servicios):
        self.servicios = servicios
        
    def getDuracion(self):
        return self.duracion
    
    def setDuracion(self, duracion):
        self.duracion = duracion
        
    def getNumeroAsistentes(self):
        return self.numeroAsistentes
    
    def setNumeroAsistentes(self, numeroAsistentes):
        self.numeroAsistentes = numeroAsistentes
        
    def getEmpleados(self):
        return self.empleados
    
    def setEmpleados(self, empleados):
        self.empleados = empleados
        
    def getFecha(self):
        return self.fecha
    
    def setFecha(self, fecha):
        self.fecha = fecha
        
    def getCodigo(self):
        return self.codigo
    
    def setCodigo(self, codigo):
        self.codigo = codigo
        
    def getValor(self):
        return self.valor
    
    def setValor(self, valor):
        self.valor = valor
        
    def getDescripcion(self):
        return self.descripcion
    
    def setDescripcion(self, descripcion):
        self.descripcion = descripcion
        
    """
    Este es un getter no tan obvio: se encarga de recorrer cada uno de los 
    servicios externos, de modo que obtiene una lista con las empresas contratadas
    para cada uno de ellos y posteriormente la retorna.
    """
    def getEmpresasContratadas(self):
        lista = []

        for servicio in self.servicios:
            lista.append(servicio.getEmpresaContratada())

        return lista

        
        